### PR TITLE
upgrade(installer): Add APM injector package installation support

### PR DESCRIPTION
**Pull request by BaptisteFoy**
### What does this PR do?
Adds the required installation commands for the APM Injector on bootstrap / experiment start / experiment stop

### Motivation
Injector bootstrap / updates

### Additional Notes
N/A

### Possible Drawbacks / Trade-offs
1. Sockets are used and configured in the agent - this is not reflected in the experiment process yet
2. There is a single file (/etc/ld.so.preload) that has to be written to by all versions - it's overridden each time.

### Describe how to test/QA your changes
(Note: requires https://github.com/DataDog/auto_inject/pull/157 & https://github.com/DataDog/auto_inject/pull/161)

1. Install the updater on a VM (you can use https://github.com/DataDog/datadog-agent/pull/24372#issuecomment-2036829660)
2. Make sure your agent configuration is correct and you're able to see it
3. Bootstrap the injector
4. Bootstrap the java tracer
5. [Install docker on your VM](https://docs.docker.com/engine/install/ubuntu/#install-using-the-repository)
6. Build [this java app](https://www.baeldung.com/dockerizing-spring-boot-application) & build [the docker image](https://github.com/eugenp/tutorials/tree/master/spring-cloud-modules/spring-cloud-docker/docker-message-server)
7. Run the docker image with `docker run -p8887:8888 message-server:latest`
8. Check that you have traces for `service:message-server`

### CODE REVIEW

File: pkg/updater/service/apm_injector.go [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+	}
+
+	err = setupLDPreload(injectorPath)
+	if err != nil {
+		return err
+	}
+
+	err = setupDockerDaemon(injectorPath)
+	if err != nil {
+		return err
```
Comment (raphaelgavache) [inline pkg/updater/service/apm_injector.go:34]: on each err, shouldn't you cleanup especially LDPreload?

Comment (BaptisteFoy) [reply pkg/updater/service/apm_injector.go:44]: Done in [fix(installer): Cleanup APM injector on setup failure](https://github.com/DataDog/datadog-agent/pull/24372/commits/ecba82511581b82fda25ef85e7aa8c4940d76b43). There may just be a side effect that if the `ld.so.preload` write fails, the uninstall may fail for the same reason (e.g. permission issues) and that will emit a log but nothing should be written, hence the _warn_

File: test/new-e2e/tests/updater/linux_test.go [hunk 2]
```diff
@@ -37,12 +38,13 @@ const (
...
 	e2e.BaseSuite[environments.Host]
 	packageManager string
+	distro         os.Descriptor
 }
 
 func runTest(t *testing.T, pkgManager string, arch os.Architecture, distro os.Descriptor) {
 	reg := regexp.MustCompile(`[^a-zA-Z0-9_\-.]`)
 	testName := reg.ReplaceAllString(distro.String()+"-"+string(arch), "_")
-	e2e.Run(t, &vmUpdaterSuite{packageManager: pkgManager}, e2e.WithProvisioner(awshost.ProvisionerNoFakeIntake(
+	e2e.Run(t, &vmUpdaterSuite{packageManager: pkgManager, distro: distro}, e2e.WithProvisioner(awshost.ProvisionerNoFakeIntake(
```
Comment (raphaelgavache) [inline test/new-e2e/tests/updater/linux_test.go:47]: get latest main to as there are new tests Redhat arm and Debian x86

File: pkg/updater/service/docker.go [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+
+const (
+	tmpDockerDaemonPath = "/tmp/daemon.json.tmp"
+	dockerDaemonPath    = "/etc/docker/daemon.json"
+)
+
+// setDockerConfig sets up the docker daemon to use the APM injector
+// even if docker isn't installed, to prepare for if it is installed
+// later
+func (a *apmInjectorInstaller) setDockerConfig() error {
```
Comment (raphaelgavache) [inline pkg/updater/service/docker.go:32]: this misses extensive local tests which you can run using temp dir.
Having local test will allow you to check all scenarios and different existing files format + faster iterations if later we break this

same for apm_inject file manipulations

**Review by raphaelgavache — APPROVED** [review]
approving so that you are not blocked with the incoming breaking changes, but verify first that you pass main test + add some more tests on a follow up

File: test/new-e2e/tests/updater/linux_test.go [hunk 4]
```diff
@@ -185,6 +187,136 @@ func (v *vmUpdaterSuite) TestPurgeAndInstallAgent() {
...
+
+	/////////////////////////////////////
+	// Check injection with a real app //
+	/////////////////////////////////////
+
+	launchJavaDockerContainer(v.T(), host)
+
+	// check "Flushed traces to the API" in trace agent logs
+	require.Eventually(v.T(), func() bool {
+		_, err := host.Execute(`cat /var/log/datadog/trace-agent.log | grep "Dropping Payload due to non-retryable error"`)
```
Comment (raphaelgavache) [inline test/new-e2e/tests/updater/linux_test.go:296]: beautiful

File: pkg/updater/service/apm_inject.go [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+	err = os.WriteFile(datadogConfigPath, content, 0644)
+	if err != nil {
+		return err
+	}
+
+	return restartTraceAgent()
+}
+
+// backupAgentConfig backs up the agent configuration
+func backupAgentConfig() error {
```
Comment (raphaelgavache) [inline pkg/updater/service/apm_inject.go:261]: you can assert that your command is correctly interpreted by the helper with these tests:
https://github.com/DataDog/datadog-agent/blob/cca8e95db5a54fdc3c73b7bed78588a12b922a15/pkg/updater/service/systemd_test.go#L50-L58

**Pull request merged**